### PR TITLE
Decouple Narayana LRA extension from RESTEasy Classic

### LINKED ISSUE #20079: Application cannot be compiled when RestEasy Reactive is used together with LRA extension
### Describe the bug

An Exception occurs when building a Quarkus application with `quarkus-resteasy-reactive` and `quarkus-narayana-lra`: 

`Failed to build quarkus application: java.lang.IllegalArgumentException: Multiple matching properties for name "security.jaxrs.deny-unannotated-endpoints" property was matched by both public boolean io.quarkus.resteasy.runtime.JaxRsSecurityConfig.denyJaxRs and public boolean io.quarkus.resteasy.reactive.common.runtime.JaxRsSecurityConfig.denyJaxRs. This is likely because you have an incompatible combination of extensions that both define the same properties (e.g. including both reactive and blocking database extensions)`

### Expected behavior

An application using `quarkus-resteasy-reactive` and `quarkus-narayana-lra` should compile and run.

### Actual behavior

The error occurs when a project is built with quarkus: build and has the following dependencies:

```xml
  <dependencies>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-resteasy-reactive</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-narayana-lra</artifactId>
    </dependency>
...
```
The error occurs not only when a rest resource is implemented, but also when the project does not contain a single line of code.

### How to Reproduce?

Reproducer:

https://github.com/haschi/lra-demo

1. start a lra-coordinator  on http://localhost:8080 (not included in reproducer)
2. run `./mvnw quarkus:build` => ok
3. switch to branch `reactive`
4. again `./mvnw quarkus:build` => error

### Output of `uname -a` or `ver`

Linux xxx 5.11.0-34-generic #36~20.04.1-Ubuntu SMP Fri Aug 27 08:06:32 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "11.0.11" 2021-04-20 OpenJDK Runtime Environment (build 11.0.11+9-Ubuntu-0ubuntu2.20.04) OpenJDK 64-Bit Server VM (build 11.0.11+9-Ubuntu-0ubuntu2.20.04, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.2.2.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.1 (05c21c65bdfed0f71a2f2ada8b84da59348c4c5d)

### Additional information

If I overlooked the fact that the LRA extension does not work with RestEasy Reactive, it would be nice to highlight 
this in the documentation.

If it is missing, it would be helpful to add it accordingly.

### PR DESCRIPTION
**Pull request by geoand**
* Decouple Narayana LRA from RESTEasy Classic (the extension now does not have a hard dependency on it, but can work with either RESTEasy Reactive or RESTEasy Classic)

* Decouple Narayana LRA extension from rest-client implementation so users can choose whether to use `quarkus-rest-client` or `quarkus-rest-client-reactive`

Both of the commits are breaking changes and since they are about the same extension, I think it's better to perform them in one go.

Fixes: #20079

### CODE REVIEW

File: docs/src/main/asciidoc/lra.adoc [hunk 1]
```diff
@@ -58,8 +58,17 @@ This will add the following to your pom.xml:
...
     </dependency>
+    <dependency>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-resteasy-jackson</artifactId>
+    </dependency>
 ----
 
+[IMPORTANT]
+====
+`quarkus-narayana-lra` can work with either `quarkus-resteasy-jackson` or `quarkus-resteasy-reactive-jackson`. Without these extensions, `quarkus-narayana-lra` has no effect
```
Comment (michalszynkiewicz) [inline docs/src/main/asciidoc/lra.adoc:69]: Wouldn't it be better if it failed with an error saying that one of these extensions is needed?

Comment (geoand) [reply docs/src/main/asciidoc/lra.adoc:69]: That's probably a good idea, let me do that.

Comment (mmusgrov) [reply docs/src/main/asciidoc/lra.adoc:69]: Agreed.
And I think the easiest thing to do for now is not support reactive.

File: integration-tests/narayana-lra/pom.xml [hunk 2]
```diff
@@ -27,6 +27,21 @@
     </dependencyManagement>
 
     <dependencies>
+<!--        <dependency>-->
```
Comment (michalszynkiewicz) [inline integration-tests/narayana-lra/pom.xml:30]: is the comment here intentional?

File: integration-tests/narayana-lra/pom.xml [hunk 3]
```diff
@@ -41,9 +56,22 @@
                 </exclusion>
             </exclusions>
         </dependency>
+<!--        <dependency>-->
```
Comment (michalszynkiewicz) [inline integration-tests/narayana-lra/pom.xml:59]: ditto

File: integration-tests/narayana-lra/pom.xml (hunk 2 above)
Comment (geoand) [reply integration-tests/narayana-lra/pom.xml]: Good catch

File: integration-tests/narayana-lra/pom.xml (hunk 3 above)
Comment (geoand) [reply integration-tests/narayana-lra/pom.xml]: Fixed

File: docs/src/main/asciidoc/lra.adoc (hunk 1 above)
Comment (geoand) [reply docs/src/main/asciidoc/lra.adoc]: Well, it comes for free, so no reason not to support it :)

**Review by xstefank — COMMENTED** [review]
Just idea. Would it be possible to default to one option classic/reactive and with command-line switch to other? 

```
./mvnw quarkus:add-extension -Dextensions="narayana-lra"
```
and

```
./mvnw quarkus:add-extension -Dextensions="narayana-lra" -Dreactive
```

File: extensions/narayana-lra/runtime/pom.xml [hunk 4]
```diff
@@ -16,11 +16,7 @@
     <dependencies>
         <dependency>
             <groupId>io.quarkus</groupId>
-            <artifactId>quarkus-resteasy-jackson</artifactId>
-        </dependency>
-        <dependency>
-            <groupId>io.quarkus</groupId>
-            <artifactId>quarkus-rest-client</artifactId>
+            <artifactId>quarkus-rest-client-reactive</artifactId>
```
Comment (michalszynkiewicz) [inline extensions/narayana-lra/runtime/pom.xml:19]: won't a dependency on quarkus-rest-client-reactive here mean that it won't work with the classic one?

**Comment by geoand** [comment]
> Just idea. Would it be possible to default to one option classic/reactive and with command-line switch to other?
> 
> ```
> ./mvnw quarkus:add-extension -Dextensions="narayana-lra"
> ```
> 
> and
> 
> ```
> ./mvnw quarkus:add-extension -Dextensions="narayana-lra" -Dreactive
> ```

@aloubyansky has an idea about how that could be done.

File: extensions/narayana-lra/runtime/pom.xml (hunk 4 above)
Comment (geoand) [reply extensions/narayana-lra/runtime/pom.xml:19]: Won't work with what?

Comment (michalszynkiewicz) [reply extensions/narayana-lra/runtime/pom.xml:19]: `rest-client-jackson`

Comment (michalszynkiewicz) [reply extensions/narayana-lra/runtime/pom.xml:19]: sorry, `quarkus-resteasy-jackson`

Comment (geoand) [reply extensions/narayana-lra/runtime/pom.xml:19]: It works just fine, that's what the integration test actually tests :)

The reactive rest client does **not** need the RESTEasy Reactive to work.

Comment (michalszynkiewicz) [reply extensions/narayana-lra/runtime/pom.xml:19]: I know it doesn't but I thought there may be some conflicts. Cool :)

**Review by michalszynkiewicz — APPROVED** [review]

File: extensions/narayana-lra/runtime/pom.xml (hunk 4 above)
Comment (geoand) [reply extensions/narayana-lra/runtime/pom.xml:19]: Nope, it's all good :)

**Comment by geoand** [comment]
I pushed a new (first) commit that will help in this an other cases as well.

**Comment by geoand** [comment]
Pushed what I hope is the final change where now the user needs to pick the rest-client implementation as well instead of the extension having a hard dependency on either of the rest client implementations Quarkus provides

**Review by aloubyansky — APPROVED** [review]

**Comment by mmusgrov** [comment]
Thanks @geoand very much appreciated.
Am I allowed to update the [quarkus blog](https://quarkus.io/blog/using-lra/) to reflect your changes?

**Comment by geoand** [comment]
You are welcome!

Yes of course. When 2.5 is released, then you can open a Pull Request against [this](https://github.com/quarkusio/quarkusio.github.io/blob/develop/_posts/2021-09-01-using-lra.adoc) to mention what changes where made in `2.5`.

**Pull request merged**